### ISSUE TITLE
Swarm becomes unstable after some time / bulk sync failed / DNS errors

### ISSUE THREAD

**Issue by scopev24** [moby/moby#37342]
**Description**

We are using a three manager swarm and a custom overlay network created with the attachable flag.
All machines run RHEL 7.4, there's no firewall in between. The nodes are called app1, app2 and app3.

There is a single swarm service running (apache webserver) which gets accesses via an nginx-proxy that runs as a local container (attached to the overlay network). Furthermore there are quite a lot of cron jobs that are started as local containers as well as some additional "services" (e.g. redis).

The swarm service runs as "global", redis runs on app1.

After initial setup the system runs smoothly. After a while (unfortunately I cannot be more precise...) the system becomes unstable. We get a lot of "unknown host" errors inside our application on app2 - e.g. when trying to access a redis container (running on app1, local container, attached to the overlay network).
I have not been able to manually reproduce these errors, but they happen very frequently. If I run my own debug container or attach to one of the running containers, I've never seen a lookup error (i.e. running a ping or a nslookup on the container name is just fine).

The log is flooded with the following errors:

```
Jun 25 16:24:09 app2 dockerd: time="2018-06-25T16:24:09.273214385+02:00" level=warning msg="bulk sync to node 0f680e5b6b13 failed: failed to send a TCP message during bulk sync: write tcp 10.105.131.70:35958->10.105.131.68:7946: write: connection reset by peer"
Jun 25 16:24:09 app2 dockerd: time="2018-06-25T16:24:09.273295194+02:00" level=error msg="periodic bulk sync failure for network kq2j6rrmpxee1mfagccvea8o9: bulk sync to node 0f680e5b6b13 failed: failed to send a TCP message during bulk sync: write tcp 10.105.131.70:35958->10.105.131.68:7946: write: connection reset by peer"
Jun 25 16:24:11 app2 dockerd: time="2018-06-25T16:24:11.182482766+02:00" level=warning msg="memberlist: failed to receive: Remote node state is larger than limit (30567266) from=10.105.131.68:42734"
Jun 25 16:24:14 app2 dockerd: time="2018-06-25T16:24:14.646395787+02:00" level=warning msg="memberlist: failed to receive: Remote node state is larger than limit (30566959) from=10.105.131.68:42742"
```

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:      18.03.1-ce
 API version:  1.37
 Go version:   go1.9.5
 Git commit:   9ee9f40
 Built:        Thu Apr 26 07:20:16 2018
 OS/Arch:      linux/amd64
 Experimental: false
 Orchestrator: swarm

Server:
 Engine:
  Version:      18.03.1-ce
  API version:  1.37 (minimum version 1.12)
  Go version:   go1.9.5
  Git commit:   9ee9f40
  Built:        Thu Apr 26 07:23:58 2018
  OS/Arch:      linux/amd64
  Experimental: false
```

**Output of `docker info`:**

```
Containers: 32
 Running: 29
 Paused: 0
 Stopped: 3
Images: 7
Server Version: 18.03.1-ce
Storage Driver: overlay2
 Backing Filesystem: xfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: 1aife2eipnelvxw2enugh4t0i
 Is Manager: true
 ClusterID: 4f6i99yx17lfmnqi7510bfizl
 Managers: 3
 Nodes: 3
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 10
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Autolock Managers: false
 Root Rotation In Progress: false
 Node Address: 10.105.131.70
 Manager Addresses:
  10.105.131.68:2377
  10.105.131.70:2377
  10.105.131.72:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 773c489c9c1b21a6d78b5c538cd395416ec50f88
runc version: 4fc53a81fb7c994640722ac585fa9ca548971871
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-693.21.1.el7.x86_64
Operating System: Red Hat Enterprise Linux Server 7.4 (Maipo)
OSType: linux
Architecture: x86_64
CPUs: 24
Total Memory: 31.12GiB
Name: app2
ID: 7CKM:M5BL:GOSN:TVCL:32YP:GVNI:BJKZ:U4XP:WR4C:KNV5:PEIO:DFDJ
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

A `docker node ls`does not show anything suspicious:

```
[nscheer@app2 ~]$ docker node ls
ID                            HOSTNAME            STATUS              AVAILABILITY        MANAGER STATUS      ENGINE VERSION
1aife2eipnelvxw2enugh4t0i *   app2                Ready               Active              Reachable           18.03.1-ce
pla8sgee6y4h12yb7ai1m898q     app3                Ready               Active              Leader              18.03.1-ce
yif5tg3mt59zn82kgmjgp26yi     app1                Ready               Active              Reachable           18.03.1-ce
```

Here are the defined networks:

```
[nscheer@app2 ~]$ docker network ls
NETWORK ID          NAME                 DRIVER              SCOPE
2e529da4ff30        bridge               bridge              local
91a3df8931f8        docker_gwbridge      bridge              local
dca56c2748e9        host                 host                local
7dhd4a592ttf        ingress              overlay             swarm
kq2j6rrmpxee        n_app_production     overlay             swarm
10d411bb9b1b        none                 null                local
```

And details of the used overlay network:

```
[nscheer@app2~]$ docker network inspect kq2j6rrmpxee
[
    {
        "Name": "n_app_production",
        "Id": "kq2j6rrmpxee1mfagccvea8o9",
        "Created": "2018-06-14T20:58:11.018335443+02:00",
        "Scope": "swarm",
        "Driver": "overlay",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": "10.0.1.0/24",
                    "Gateway": "10.0.1.1"
                }
            ]
        },
        "Internal": false,
        "Attachable": true,
        "Ingress": false,
        "ConfigFrom": {
            "Network": ""
        },
        "ConfigOnly": false,
        "Containers": {
            "4491bd2f0939506ac9591a1cda22bd67c5646a093cd3b3cdf727467d5a69e787": {
                "Name": "app-production-bfp-1.0.5-cron-10672-2018-06-25_16.00.02",
                "EndpointID": "03a8abadaf712bce175fef99e97fc8a2ea9c7a0a9d3adca70870e7c037a5b78f",
                "MacAddress": "02:42:0a:00:01:87",
                "IPv4Address": "10.0.1.135/24",
                "IPv6Address": ""
            },
            "520431731b4a99fbcead85170f191887f60ed417a78e74e6b377933f14973f94": {
                "Name": "app-production-bfp-1.0.5-cron-33725-2018-06-25_16.15.01",
                "EndpointID": "fc770fe143c113853943ba9a1d871ea546576cf8703635cb35fe4422d3091459",
                "MacAddress": "02:42:0a:00:01:49",
                "IPv4Address": "10.0.1.73/24",
                "IPv6Address": ""
            },

            "7b249f84980e4cac3e0fc18e839098bbfeaac8830410fc219464b84b5a615652": {
                "Name": "app-production-bfp-1.0.5-cron-11444-2018-06-14_21.54.01",
                "EndpointID": "b31d41ff3e21204794ecc0e6cb77f6c80a527f626774cf714c329c6d75094b8b",
                "MacAddress": "02:42:0a:00:01:6c",
                "IPv4Address": "10.0.1.108/24",
                "IPv6Address": ""
            },
            "a2ec03e23e79e81990ab7b500ab42620902f750bf0c2ff928a75098445806ae7": {
                "Name": "app-nginx-proxy",
                "EndpointID": "b4f711b5c4498205f7dc867a3cec9faeaf38290cd43ce8e67b4fe322d9b7466b",
                "MacAddress": "02:42:0a:00:01:0f",
                "IPv4Address": "10.0.1.15/24",
                "IPv6Address": ""
            },
            "a40fbc5651ada0bba2a7bfb46ae56300e06ead95cc462ea20e06328e5a0d7eac": {
                "Name": "app-production-bfp_web.1aife2eipnelvxw2enugh4t0i.2v38czull38w85uno7g02580t",
                "EndpointID": "a4781faf4cf75b29898ba63266ee19f5045deb48528711bdb35b967c5fb73e66",
                "MacAddress": "02:42:0a:00:01:ca",
                "IPv4Address": "10.0.1.202/24",
                "IPv6Address": ""
            },
            "b3f48b0890a0fada8a0950477ddad4e4aed7a6522d24ea591c5feead1cf7befe": {
                "Name": "app-production-bfp-1.0.5-cron-14869-2018-06-14_21.55.01",
                "EndpointID": "b5acaa446d176669d6b5c1f2ad6115ce7706ecc57051eb0fc88b4724d4377286",
                "MacAddress": "02:42:0a:00:01:73",
                "IPv4Address": "10.0.1.115/24",
                "IPv6Address": ""
            }
        },
        "Options": {
            "com.docker.network.driver.overlay.vxlanid_list": "4098"
        },
        "Labels": {},
        "Peers": [
            {
                "Name": "c1e982bb7334",
                "IP": "10.105.131.70"
            },
            {
                "Name": "0f680e5b6b13",
                "IP": "10.105.131.68"
            },
            {
                "Name": "f8d6a591692e",
                "IP": "10.105.131.72"
            }
        ]
    }
]
```

Draining and removing single nodes in order to restart the docker daemon does not resolve this issue.
The only thing that has helped so far was to stop all containers and services, delete and recreate the overlay network and starting everything again. Of course, on a production system like this, this can only be done as a last resort.

I'm happy to provide any additional details - unfortunately I do not know where to start my search.

Thanks for your help!

Greetings

Nico

**Comment by scopev24** [comment]
Hi again,

seems like the dns issues correlate with load spikes - I'm currently trying to reproduce these on a test system.
Nevertheless, can anybody be so kind to elaborate on the docker errors I'm seeing? Unfortunately I'm unable to find any information regarding "bulk sync" failures and "Remote node state is larger than limit".

Thanks for your help!

**Comment by scopev24** [comment]
Hi again,

I managed to re-setup the whole swarm - I used 18.06.0-rc1.
I was able to reproduce the DNS issues - they occured under high load and effectively were "just" dns timeouts. I don't know, why the docker daemon would take so long (> 4 seconds) to answer a dns query. The case could be reproduced just by calling nslookup in a loop and watch it fail every 5 minutes (when many containers are spun up through cron).
Release 18.06.0-rc1 did not show that behaviour any more - so I stopped my research.

The aforementioned errors regarding bulk sync have not re-occurred so far - That might have been a inconsistency due to some network re-configuration on the host.

Closing this issue.
